geode-core/src/main/java/org/apache/geode/distributed/internal/InternalLocator.java

```diff
@@ -773,15 +773,19 @@ private void startCache(DistributedSystem system) throws IOException {
     startClusterManagementService();
   }
 
-  private void startClusterManagementService() throws IOException {
+  @VisibleForTesting
+  void startClusterManagementService() throws IOException {
     startConfigurationPersistenceService();
 
-    if (internalCache == null) {
+    InternalCache myCache = this.internalCache;
+    InternalLocator currentLocator = InternalLocator.locator;
+
+    if (myCache == null || currentLocator == null) {
```

Comment:
If this expression is true it seems like we just silently returned without starting all of the locator's services. I have seen runs (recently) in which this happens because the InternalDistributedSystem does a concurrent InternalLocator.stop while another thread is executing the restart of this locator. So before returning here should this locator be marked in some way as not having been started? I think the thread that waits for this thread (the one executing startClusterManagementService) just joins and then says we are started.